services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/device/AutoProvisioner.java

```diff
@@ -0,0 +1,386 @@
+                });
+            }
+        }
+
+        return connectionAttempt.mapEmpty();
+    }
+
+    @Override
+    public final Future<Void> stop() {
+        final Promise<Void> result = Promise.promise();
+        downstreamSenderFactory.disconnect(result);
+        return result.future();
+    }
+
+    /**
+     * Sets the vert.x instance.
+     *
+     * @param vertx The vert.x instance.
+     * @throws NullPointerException if vert.x is {@code null}.
+     */
+    @Autowired
+    public final void setVertx(final Vertx vertx) {
+        this.vertx = Objects.requireNonNull(vertx);
+    }
+
+    /**
+     * Sets the factory to use for creating a client for the AMQP Messaging Network.
+     *
+     * @param factory The factory.
+     * @throws NullPointerException if the factory is {@code null}.
+     */
+    @Autowired
+    public final void setDownstreamSenderFactory(final DownstreamSenderFactory factory) {
+        this.downstreamSenderFactory = Objects.requireNonNull(factory);
+    }
+
+    /**
+     * Sets the service to use for checking existence of tenants.
+     * <p>
+     * If not set, tenant existence will not be verified.
+     *
+     * @param tenantInformationService The tenant information service.
+     * @throws NullPointerException if service is {@code null};
+     */
+    @Autowired(required = false)
+    public final void setTenantInformationService(final TenantInformationService tenantInformationService) {
+        this.tenantInformationService = Objects.requireNonNull(tenantInformationService);
+        LOG.info("using {}", tenantInformationService);
+    }
+
+    /**
+     * Sets the OpenTracing {@code Tracer} to use for tracking the processing
+     * of messages published by devices across Hono's components.
+     * <p>
+     * If not set explicitly, the {@code NoopTracer} from OpenTracing will
+     * be used.
+     *
+     * @param opentracingTracer The tracer.
+     * @throws NullPointerException if the opentracingTracer is {@code null}.
+     */
+    @Autowired(required = false)
+    public final void setTracer(final Tracer opentracingTracer) {
+        LOG.info("using OpenTracing Tracer implementation [{}]", opentracingTracer.getClass().getName());
+        this.tracer = Objects.requireNonNull(opentracingTracer);
+    }
+
+    /**
+     * Sets the configuration to use for auto-provisioning.
+     *
+     * @param config The configuration to set.
+     * @throws NullPointerException if the config is {@code null}.
+     */
+    @Autowired
+    public void setConfig(final AutoProvisionerConfigProperties config) {
+        this.config = Objects.requireNonNull(config);
+    }
+
+    private Future<ProtonDelivery> sendAutoProvisioningEvent(
+            final String tenantId,
+            final String deviceId,
+            final String gatewayId,
+            final Span span) {
+
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(deviceId);
```

Comment:
Also the `span` should be null checked, otherwise `span.context()` will throw error.